### ISSUE TITLE
Provide units that remain “Starting” until explicitly marked as having started

### ISSUE THREAD

**Issue by DemiMarie** [systemd/systemd#23732]
**Is your feature request related to a problem? Please describe.**
There are times I need to block the execution of a service until some external, asynchronous event happens.  Right now, this requires either blocking in `ExecStartPre=` (such as by reading a FIFO with no writers) or creating a fake device of some sort so that a `.device` unit can be used.  Both are ugly.

**Describe the solution you'd like**
I would like to to be able create a unit (name TBD, perhaps `.event`?) that is marked as “starting” whenever something requests it to be started via a dependency, but will not be considered “running” until a specific message is sent to systemd by an arbitrary process, which may not be part of the unit.

**Describe alternatives you've considered**
See above.

**The systemd version you checked that didn't have the feature you are asking for**
249?  No entries in NEWS or issues that I could find, anyway.

**Comment by dtardon** [comment]
> **Describe the solution you'd like** I would like to to be able create a unit (name TBD, perhaps `.event`?) that is marked as “starting” whenever something requests it to be started via a dependency, but will not be considered “running” until a specific message is sent to systemd via D-Bus.

`Type=notify` does this, just not via D-Bus. (And `Type=dbus` does this too, sort of. The "specific message" is sent by acquiring the configured bus name.)

**Comment by DemiMarie** [comment]
> > **Describe the solution you'd like** I would like to to be able create a unit (name TBD, perhaps `.event`?) that is marked as “starting” whenever something requests it to be started via a dependency, but will not be considered “running” until a specific message is sent to systemd via D-Bus.
> 
> `Type=notify` does this, just not via D-Bus. (And `Type=dbus` does this too, sort of. The "specific message" is sent by acquiring the configured bus name.)

`Type=notify` has the process started by the unit notifies systemd when it is ready.  The type of unit I am referring to has no processes.

**Comment by dtardon** [comment]
> `Type=notify` has the process started by the unit notifies systemd when it is ready.

Exactly. The process is started, it waits for the external event, and only then notifies systemd it's ready. Isn't this what you wanted? If not, then I completely misunderstood the description...

**Comment by DemiMarie** [comment]
> > `Type=notify` has the process started by the unit notifies systemd when it is ready.
> 
> Exactly. The process is started, it waits for the external event, and only then notifies systemd it's ready. Isn't this what you wanted? If not, then I completely misunderstood the description...

In my case, the process is *not* started by the unit.

**Comment by poettering** [comment]
Why not follow the model of "systemd-network-wait-online.service" or "systemd-time-wait-sync.service"? i.e. have service that waits for the event?

**Comment by DemiMarie** [comment]
> Why not follow the model of "systemd-network-wait-online.service" or "systemd-time-wait-sync.service"? i.e. have service that waits for the event?

It requires writing custom code for each such service, so it is not suitable for one-off scripts (very common).

**Comment by poettering** [comment]
i mean either way you need to write custom code, because something needs to tell systemd anyway...

**Comment by DemiMarie** [comment]
> i mean either way you need to write custom code, because something needs to tell systemd anyway...

That something would typically be a script that calls `systemctl`.

@poettering: Is @boucman’s solution reliable?  If so, it could be added to the systemd documentation.  @boucman: have you had situations where using a target would not have sufficed?

**Comment by poettering** [comment]
No, @boucman's suggestion does not work. order is only enforced between jobs actually queued. If you never queue the target unit, the ordering towards it has zero effect. But if you do queue it, then it will just be activated, because systemd can just do that. You are asking for something where systemd cannot in fact just activate the unit when it wants to, but wants you to do that.
